components/public-api/gitpod/v1/configuration.proto

```diff
@@ -97,6 +98,14 @@ message UpdateConfigurationRequest {
   }
   message WorkspaceSettings {
     optional string workspace_class = 1;
+
+    // restricted_workspace_classes specifies the workspace classes that are NOT allowed to be used in this configuration.
+    // If empty, all workspace classes are allowed.
+    // Only update if update_restricted_workspace_classes is true.
+    repeated string restricted_workspace_classes = 2;
+
+    // update_restricted_workspace_classes specifies whether restricted_workspace_classes should be updated.
+    optional bool update_restricted_workspace_classes = 3;
```

Comment:
I am not super happy about having fields that depend on each other, since it makes the API really non-intuitive and hard to use without documentation. But if it's difficult to achieve any other way, let's use it. 

(Picking up on the thread in https://github.com/gitpod-io/gitpod/pull/19410#discussion_r1504690850)

cc @jeanp413 @mustard-mh

Reply:
@filiptronicek  It's a suggestion from ChatGPT, we have other options like FieldMask, but for user's use case, it will be more complex

Reply:
FieldMask will make request like, (fake code)

```ts
new UpdateConfigurationSettingsRequest({
   ...
   workspaceSettings: {
     restrictedWorksapceClasses: ["g1", "g2"],
   },
   updateMask: new FieldMask({
     paths: ["workspaceSettings.restrictedWorksapceClasses"]
   })
})
```

Reply:
Doc need to be

```proto
message UpdateConfigurationSettingsRequest{
  // update_mask sets field user want to update
  // affected fields and available values:
  // `workspaceSettings.restrictedWorksapceClasses`: update `restricted_workspace_classes` field
  FieldMask update_mask = 1;
}
```

Reply:
> I am not super happy about having fields that depend on each other

We have no choice, that's the limit of `repeated` fields

Reply:
> since it makes the API really non-intuitive and hard to use without documentation.

@filiptronicek  If the `restricted_workspace_classes` is set (check length > 0) without `update_restricted_workspace_classes` we will throw errors

Without documentation, it's hard to use any APIs

<img width="769" alt="image" src="https://github.com/gitpod-io/gitpod/assets/20944364/fc5dab18-219d-4d1d-872e-2fa33c0504e4">

Reply:
> since it makes the API really non-intuitive and hard to use without documentation

Additional example for this comment, we have ordering in PAPI, only read through document could know which field is working. So they are the same

Reply:
Thanks for the explainer and teaching me about field masks. Haven't heard about them before!

Reply:
I don't know it too 🤣, we have some discussion around update `repeated` fields in [this thread (internal)](https://gitpod.slack.com/archives/C0463A2KQUU/p1702382520145059) and about [FieldMask itself (internal)](https://gitpod.slack.com/archives/C0463A2KQUU/p1699617451051229)

@filiptronicek If you're interested with